File: Scheduler/Samples/Scheduler.MVC3/Controllers/ServerSideFilteringController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

using DHTMLX.Scheduler;
using DHTMLX.Scheduler.Data;
using DHTMLX.Scheduler.Controls;
using DHTMLX.Common;
using SchedulerTest.Models;

namespace SchedulerTest.Controllers
{
    public class ServerSideFilteringController : Controller
    {
        //
        // GET: /ServerSideFiltering/
        public class SchedulerFilterModel
        {
            public DHXScheduler Scheduler { get; set; }
            public IEnumerable<Room> Rooms { get; set; }
         
            public SchedulerFilterModel(DHXScheduler sched, IEnumerable<Room> rooms)
            {
                this.Scheduler = sched;
                this.Rooms = rooms;
              
            }
        }

        public ActionResult Index(FormCollection data)
        {

            var sched = new DHXScheduler(this);

            sched.Extensions.Add(SchedulerExtensions.Extension.Cookie);

            var rooms = new DHXSchedulerDataContext().Rooms.ToList();
                    
            int selectedRoom;
            if (this.Request.QueryString["filter"] != null)
            {
                // parameters will be added to data url
                sched.Data.Loader.AddParameters(this.Request.QueryString);
                selectedRoom = int.Parse(this.Request.QueryString["rooms"]);
            }
            else
            {
                selectedRoom = rooms.First().key;               
            }


            var unit = new UnitsView("rooms", "room_id");
            unit.Label = "Rooms";
            unit.AddOptions(rooms);
            sched.Views.Add(unit);


            sched.Lightbox.AddDefaults();
            var select = new LightboxSelect("room_id", "Room");
            select.AddOptions(rooms);
            sched.Lightbox.Add(select); 


            sched.LoadData = true;
            sched.EnableDataprocessor = true;
            sched.InitialView = unit.Name;
            ViewData["rooms"] = selectedRoom;
            sched.InitialDate = new DateTime(2011, 9, 7);
            return View(new SchedulerFilterModel(sched, rooms));

        }
        public ContentResult Data()
        {
            var dc = new DHXSchedulerDataContext();

            IEnumerable<Event> dataset;

            if (this.Request.QueryString["rooms"] == null)
                dataset = dc.Events;
            else{
                var current_room = int.Parse(this.Request.QueryString["rooms"]);
                dataset = from ev in dc.Events where ev.room_id == current_room select ev;
            }


            var data = new SchedulerAjaxData(dataset);

   
            return (data);
        }

        public ContentResult Save(int? id, FormCollection actionValues)
        {

            var action = new DataAction(actionValues);

            DHXSchedulerDataContext data = new DHXSchedulerDataContext();
            var changedEvent = DHXEventsHelper.Bind<Event>(actionValues);
            try
            {
                switch (action.Type)
                {
                    case DataActionTypes.Insert:
                        data.Events.InsertOnSubmit(changedEvent);
                        break;
                    case DataActionTypes.Delete:
                        changedEvent = data.Events.SingleOrDefault(ev => ev.id == action.SourceId);
                        data.Events.DeleteOnSubmit(changedEvent);
                        break;
                    default:// "update"                          
                        var eventToUpdate = data.Events.SingleOrDefault(ev => ev.id == action.SourceId);
                        DHXEventsHelper.Update(eventToUpdate, changedEvent, new List<string>() { "id" });
                        break;
                }
                data.SubmitChanges();
                action.TargetId = changedEvent.id;
            }
            catch (Exception a)
            {
                action.Type = DataActionTypes.Error;
            }

            return (new AjaxSaveResponse(action));
        }

    }
}

File: Scheduler/Samples/Scheduler.MVC3.VB/Controllers/ServerSideFilteringController.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Web.Mvc

Imports DHTMLX.Scheduler
Imports DHTMLX.Scheduler.Data
Imports DHTMLX.Scheduler.Controls
Imports DHTMLX.Common


Namespace SchedulerTest.Controllers
	Public Class ServerSideFilteringController
		Inherits Controller
		'
		' GET: /ServerSideFiltering/
		Public Class SchedulerFilterModel
			Public Property Scheduler() As DHXScheduler
				Get
					Return m_Scheduler
				End Get
				Set
					m_Scheduler = Value
				End Set
			End Property
			Private m_Scheduler As DHXScheduler
			Public Property Rooms() As IEnumerable(Of Room)
				Get
					Return m_Rooms
				End Get
				Set
					m_Rooms = Value
				End Set
			End Property
			Private m_Rooms As IEnumerable(Of Room)

			Public Sub New(sched As DHXScheduler, rooms As IEnumerable(Of Room))
				Me.Scheduler = sched

				Me.Rooms = rooms
			End Sub
		End Class

		Public Function Index(data As FormCollection) As ActionResult

			Dim sched = New DHXScheduler(Me)

			sched.Extensions.Add(SchedulerExtensions.Extension.Cookie)

			Dim rooms = New DHXSchedulerDataContext().Rooms.ToList()

			Dim selectedRoom As Integer
			If Me.Request.QueryString("filter") IsNot Nothing Then
				' parameters will be added to data url
				sched.Data.Loader.AddParameters(Me.Request.QueryString)
				selectedRoom = Integer.Parse(Me.Request.QueryString("rooms"))
			Else
				selectedRoom = rooms.First().key
			End If


			Dim unit = New UnitsView("rooms", "room_id")
			unit.Label = "Rooms"
			unit.AddOptions(rooms)
			sched.Views.Add(unit)


			sched.Lightbox.AddDefaults()
			Dim [select] = New LightboxSelect("room_id", "Room")
			[select].AddOptions(rooms)
			sched.Lightbox.Add([select])


			sched.LoadData = True
			sched.EnableDataprocessor = True
			sched.InitialView = unit.Name
			ViewData("rooms") = selectedRoom
			sched.InitialDate = New DateTime(2011, 9, 7)
			Return View(New SchedulerFilterModel(sched, rooms))

		End Function
		Public Function Data() As ContentResult
			Dim dc = New DHXSchedulerDataContext()

			Dim dataset As IEnumerable(Of [Event])

			If Me.Request.QueryString("rooms") Is Nothing Then
				dataset = dc.Events
			Else
				Dim current_room = Integer.Parse(Me.Request.QueryString("rooms"))
                dataset = From ev In dc.Events Where ev.room_id = current_room
			End If


			Dim data__1 = New SchedulerAjaxData(dataset)


			Return (data__1)
		End Function

		Public Function Save(id As System.Nullable(Of Integer), actionValues As FormCollection) As ContentResult

			Dim action = New DataAction(actionValues)

			Dim data As New DHXSchedulerDataContext()
			Dim changedEvent = DirectCast(DHXEventsHelper.Bind(GetType([Event]), actionValues), [Event])
			Try
				Select Case action.Type
					Case DataActionTypes.Insert
						data.Events.InsertOnSubmit(changedEvent)
						Exit Select
					Case DataActionTypes.Delete
						changedEvent = data.Events.SingleOrDefault(Function(ev) ev.id = action.SourceId)
						data.Events.DeleteOnSubmit(changedEvent)
						Exit Select
					Case Else
						' "update"                          
						Dim eventToUpdate = data.Events.SingleOrDefault(Function(ev) ev.id = action.SourceId)
						DHXEventsHelper.Update(eventToUpdate, changedEvent, New List(Of String)() From { _
							"id" _
						})
						Exit Select
				End Select
				data.SubmitChanges()
				action.TargetId = changedEvent.id
			Catch a As Exception
				action.Type = DataActionTypes.[Error]
			End Try

			Return (New AjaxSaveResponse(action))
		End Function

	End Class
End Namespace
